### PR TITLE
feat(testing): allow to set screenshot timeout option in Jest v28+

### PR DESCRIPTION
**Pull request by christian-bromann**
## What is the current behavior?
When we added support for Jest 28+, we assumed the usage of Jest Circus, the new test runtime for the Jest testing library. With Jest Circus, we lost the ability to configure the timeout related to screenshots in the Stencil Test Runner. This effort adds this support back in.

## What is the new behavior?
Users can define a `screenshotTimeout` option within the Stencil testing configuration which is then passed into the worker.

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

## Testing

To test this new option I created a reproduction case that requires some manual tweaks as I wasn't able to figure out how to create a test that exceeds 2.5s when comparing a screenshot. I assume the user who reported this does some large scale screenshot comparisons.

In order to validate this option:
- clone the repro case: `git clone git@github.com:christian-bromann/stencil-screenshotTimeout-repro.git`
- `cd stencil-screenshotTimeout-repro`
- create a baseline `npm run test`
- delay the screenshot comparison by tweaking pixelmatch in `node_modules/@stencil/core/screenshot/pixel-match.js` and add the following change in line `2311`:

```patch
- function getMismatchedPixels(pixelMatchInput) {
+ async function getMismatchedPixels(pixelMatchInput) {
+   await new Promise((resolve) => setTimeout(resolve, 5000));
```

- next, change the component applying this change:

```patch
diff --git a/src/components/my-component/my-component.e2e.ts b/src/components/my-component/my-component.e2e.ts
index 2597d36..78cf174 100644
--- a/src/components/my-component/my-component.e2e.ts
+++ b/src/components/my-component/my-component.e2e.ts
@@ -25,9 +25,9 @@ describe('my-component', () => {
     await page.waitForChanges();
     expect(element.textContent).toEqual(`Hello, World! I'm James Quincy`);
 
-    component.setProperty('middle', 'Earl');
+    component.setProperty('middle', 'Earl!!');
     await page.waitForChanges();
-    expect(element.textContent).toEqual(`Hello, World! I'm James Earl Quincy`);
+    expect(element.textContent).toEqual(`Hello, World! I'm James Earl!! Quincy`);
 
     const results = await page.compareScreenshot('My Component (...is beautiful. Look at it!)', {fullPage: false});
     expect(results).toMatchScreenshot({ allowableMismatchedPixels: 100 })
```

- now run the test again `npm test`
- you should see an error saying `thrown: "getMismatchedPixels timeout: 2500ms"`
- increase the screenshot timeout by applying this change:

```patch
diff --git a/stencil.config.ts b/stencil.config.ts
index 50db729..52c8bb5 100644
--- a/stencil.config.ts
+++ b/stencil.config.ts
@@ -20,5 +20,6 @@ export const config: Config = {
   ],
   testing: {
     browserHeadless: "new",
+    screenshotTimeout: 6000,
   },
 };
```

- now run again and you should see a mismatch error instead of a timeout

## Other information

n/a

### CODE REVIEW

**Comment by christian-bromann** [comment]
Created a dev build `4.13.0-dev.1710957581.b36e157` which passed the [Framework pipeline](https://github.com/ionic-team/ionic-framework/actions/runs/8363954634) ✅

**Comment by christian-bromann** [comment]
I updated the PR template and added a repro case. Let me know if this works or if there is a better way to repro this more "realistic".

File: src/declarations/stencil-public-compiler.ts [hunk 1]
```diff
@@ -1984,6 +1984,11 @@ export interface TestingConfig extends JestConfig {
    */
   screenshotConnector?: string;
 
+  /**
+   * Some JSDoc, describing this field
```
Comment (rwaskiewicz) [inline src/declarations/stencil-public-compiler.ts:1988]: Can we put something more descriptive here please 🤣

File: src/screenshot/screenshot-compare.ts [hunk 2]
```diff
@@ -134,9 +140,11 @@ async function getMismatchedPixels(pixelmatchModulePath: string, pixelMatchInput
      * The '2500' value that we default to is the value of `jasmine.DEFAULT_TIMEOUT_INTERVAL` (5000) divided by 2.
      */
     const timeout =
-      typeof jasmine !== 'undefined' && jasmine.DEFAULT_TIMEOUT_INTERVAL
-        ? jasmine.DEFAULT_TIMEOUT_INTERVAL * 0.5
-        : 2500;
+      screenshotTimeoutMs !== null
```
Comment (rwaskiewicz) [inline src/screenshot/screenshot-compare.ts:143]: Can we update the JSDoc for this logic, now that we don't just set the value to 2500?

**Review by rwaskiewicz — CHANGES_REQUESTED** [review]
All in all looks good!

I had a couple minor asks in the code itself. Other than that, can you add some documentation to the Stencil site regarding this new flag?

Thanks!

**Comment by christian-bromann** [comment]
Good catch @rwaskiewicz , updated docs.

**Comment by rwaskiewicz** [comment]
@christian-bromann Changes here look good! Only thing missing from my [original feedback](https://github.com/ionic-team/stencil/pull/5537#pullrequestreview-1954634201) seems to be a PR for the Stencil Site:

> Other than that, can you add some documentation to the Stencil site regarding this new flag?

**Comment by christian-bromann** [comment]
@rwaskiewicz oh I missed that, added a PR https://github.com/ionic-team/stencil-site/pull/1382

**Review by rwaskiewicz — APPROVED** [review]

**Pull request merged**